### ISSUE TITLE
Latest ipykernel breaks jupyter-ui-poll

### ISSUE THREAD

**Issue by AaronWatters** [Kirill888/jupyter-ui-poll#13]
The polling mechanism no longer works under the latest release of ipykernel from pip.

Here is a screenshot

![ui-poll-issue](https://user-images.githubusercontent.com/13857252/131500779-29be45e5-24f9-45c9-a2bb-7b4d4ffd36a5.png)

The most immediate problem is that do_one_iteration is now an async method.

https://github.com/ipython/ipykernel/blob/master/ipykernel/kernelbase.py#L421

However, ipykernel seems to have been heavily rewritten, so there may be other issues too.

**Comment by AaronWatters** [comment]
I may have come up with a work around (which I desperatetly need)
based on your Example using run_ui_poll_loop:

```python
from IPython import get_ipython
from tornado.queues import QueueEmpty

async def run_poll_loop(callback, sleep=0.02):
    shell = get_ipython()
    kernel = shell.kernel
    test = None
    while test is None:
        try:
            await kernel.do_one_iteration()
        except QueueEmpty:
            pass
        time.sleep(sleep)
        test = callback()
    return test

t0 = time.time()
xx = ['-_-', 'o_o']

def on_poll():
    """ This is called repeatedly by run_ui_poll_loop
    
        Return None if condition hasn't been met yet
        
        Return some result once done, in this example result 
        is a number of seconds it took to press the button 10 times.
    """
    if int(btn.description) < 10:
        print(xx[0], end='\r', flush=True)
        xx[:] = xx[::-1]
        return None  # Continue polling
    
    # Terminate polling and return final result
    return time.time() - t0

btn = test_button()
print('Press button 10 times')
display(btn)

#dt = run_ui_poll_loop(on_poll, 1/15)
async def go():
    global dt
    dt = await run_poll_loop(on_poll, 1/15)

await go()
print('._.')         # This should display the text in the output of this cell
n_times = '10 times' # To verify that the rest of this cell executes before executing cells below
```

I'm going to try using this as a fallback for `run_ui_poll_loop` whenever
`inspect.iscoroutinefunction(kernelbase.Kernel.do_one_iteration)` evaluates to `True`.

Unfortunately it's not a plug compatible replacement because you need to add the
`await` keyword...

**Comment by AaronWatters** [comment]
Unfortunately the above doesn't really work because it allows the next notebook cell to execute.

**Comment by Kirill888** [comment]
Thanks for the report @AaronWatters , it was bound to happen one day, I'll have a look what has changed on IPython side, do you know exactly what version of IPython broke the behaviour? by latest do you mean this one https://pypi.org/project/ipython/7.27.0/

**Comment by AaronWatters** [comment]
Here is the "About" information on the version I have exhibiting the problem:

About Jupyter Notebook
Server Information:
You are using Jupyter notebook.

The version of the notebook server is: 6.4.3
The server is running on this version of Python:
Python 3.6.8 |Anaconda, Inc.| (default, Dec 29 2018, 19:04:46) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]

Current Kernel Information:
Python 3.8.11 (default, Aug  6 2021, 08:56:27) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.26.0 -- An enhanced Interactive Python. Type '?' for help.

**Comment by AaronWatters** [comment]
Sorry -- what you want is ipykernel._version.version_info?

(6, 3, 0)

**Comment by Kirill888** [comment]
I have reproduced this locally, basically `kernel.do_one_iteration` is  now an async function (not sure from what version of `ipython`), it was not before. Converting whole thing to async would be one way to handle it, using a thread would be another. I had a quick poke at it, but haven't found any easy fixes yet, will need to investigate more.

**Comment by AaronWatters** [comment]
One problem that is hard to get around is "await anything" not in a sub-task blocks all communication with widgets until the await is completed -- unless you add do_one_iteration somewhere which messes up stdio and other internal state, I think.

Right now I'm heading towards a solution like this:

- Do all widget polling/pulling/pushing in a sub-task (not at the ipython top level task).
- Warn the user to only do "await x" in a sub task -- never at the top level.
- Raise an exception if the user tries to get a value not yet populated by a sub task like NotReadyError("please wait...")

I'm thinking there is no good way to wait for a widget operation in the main task in the current implementation.

**Comment by Kirill888** [comment]
have some prototype "working", 
easiest way is to switch the whole lot to async

```python
from tornado.queues import QueueEmpty
import asyncio

async def flush_proc(kernel):
    current_parent = (kernel._parent_ident, kernel.get_parent())
    while True:
        try:
            await kernel.do_one_iteration()
        except QueueEmpty as e:
            return
        finally:
            kernel.set_parent(*current_parent)

shell = get_ipython()           
display(test_button())        
for x in range(100):
    print('.', end='')
    await flush_proc(shell.kernel)
    await asyncio.sleep(0.1)

print("\nAll done")
```

this deals with "IO going random places", but not with "Run All" (yet)

**Comment by AaronWatters** [comment]
Unless you have more tricks up your sleeve this still results in very confusing behavior.
In particular other cells can execute while the "await" cell is not finished.  See the following screenshot

![fix_issue](https://user-images.githubusercontent.com/13857252/131879212-5d231704-9298-480b-b33f-ec46a859bf4b.png)

Note that "cell 3" starts after "cell 4" starts and it executes while "cell 4" is still not finished.

This means that my users will try to use values in later cells which have not arrived yet (unless it's fixed).

Honestly, Jupyter is already confusing enough...

**Comment by Kirill888** [comment]
I have a working prototype:

https://gist.github.com/Kirill888/30bbbd3444ef05cb03b79893fb382e54

basically from ipykernel 6, things are now async. This changes how one drives event processing, and how to replay code events.

I'll probably make a new release this weekend. Most likely I'll just make it require `ipykernel>=6`, or I might try supporting `ipykernel=5.*` as well somehow. I guess could keep current code for "sync" case and have new async code then switch at runtime based on version of a kernel running.

**Comment by AaronWatters** [comment]
Very nice!

I just tried it and it seems to work as advertised.

Maybe you can test for the newer kernel implementation like this:

```Python

import asyncio
from IPython import get_ipython
import inspect

def test_kernel_is_async():
    "True iff the kernel has been ported to use async primatives (somewhere around release 6?)"
    shell = get_ipython()
    kernel = shell.kernel
    return inspect.iscoroutinefunction(kernel.do_one_iteration)

```

and fall back to the current implementation if `test_kernel_is_async()` is `False`?

**Comment by Kirill888** [comment]
It's a bit more complicated than that, test above will tell me whether kernel is async or sync, and yes I can return `async_poll` or `sync_poll` based on that, but that does not make for a good UX as user won't know whether they should `await poll()` or just `poll()`, and also breaks other functions.

One option is to do something like:

```python
with ui_events(async=True):
    await poll()
    ...
with ui_events(async=False):
    poll()
    ...
```
Then at least user knows what to expect, but this requires doing something like `async_to_sync` and executing kernel events in a separate thread and separate async loop for `async=False` case, which might work fine or might not if there are some assumptions about the same `asyncio.loop` being used for all the messages.

I think the way forward is to provide an async version of the api which will be supported on ipykernel 6+, with optional sync adaptor which will be more likely to cause trouble. Recommendation will be use async version for best performance and robustness.

Basically when `ipykernel` runs a cell it decides whether to do in async fashion or sync fashion based on code about to be executed. When it decides to run in sync mode, you can not process any events on the main thread as loop is busy executing your code and won't process any more events until you are done, unless you run those in a separate thread in a different event loop, which I'm yet to try.

**Comment by AaronWatters** [comment]
Unfortunately you are probably right to force an ipykernel upgrade instead of supporting both modes.
My initial inclination was to use a switch and only introduce the "async" keyword to the API.
Forcing an upgrade should only be an issue if the newer versions of ipykernel also breaks other things.

Some comments:

1) I think you need to update the readme to include the async keyword in the example code where needed.
2) I'm not sure you need all the stuff you list in the requirements file, but maybe I'm missing something.

Thanks for fixing this.  I will re-release my libraries that depend on this module or let you know if I can't get them
to work.

**Comment by Kirill888** [comment]
Honestly I'm more concerned about switching to async only mode, I think I can add synchronous mode, but this will require launching a separate thread and processing ui events in a separate thread, which might be problematic in some cases.

I'm thinking of doing something along the lines of:

```python
with ui_thread():
    long_running_sync_task()

async with ui_thread():
    await long_running_async_task()
```

Here `ui_thread()` would return an object that implements both `__enter__` and `__aenter__` so same function can be used in both sync and async case. In synchronous case thread will be launched that will run a separate async loop and drive `ipykernel` event processing, in async case it will just schedule a repeating timer on the main loop that drives event processing.

@AaronWatters how are you using this by the way? Is it just in notebooks or do you wrap it in some library function/class.

**Comment by AaronWatters** [comment]
Why can't you just make the old interface async (even though it is not) and only use it if you detect it will work?
Then you could support both <6 and >6 ipykernel using the same top level API. No?

I'm using it in libraries I wrote in various places.  Here are some examples (non-exhaustive)

https://github.com/AaronWatters/jp_proxy_widget/blob/master/jp_proxy_widget/proxy_widget.py#L778

https://github.com/AaronWatters/jp_doodle/blob/master/jp_doodle/dual_canvas.py#L553

https://github.com/AaronWatters/jp_doodle/blob/master/jp_doodle/auto_capture.py#L27

https://github.com/AaronWatters/feedWebGL2/blob/master/feedWebGL2/volume.py#L237

The library has been very useful and it broke a lot of things when it stopped working  (not your fault of course).

**Comment by Kirill888** [comment]
I can certainly keep `run_ui_poll_loop` the same once I have synchronous interface ready, or maybe make it dual stack based on whether passed in function is a coroutine or normal function.

**Comment by AaronWatters** [comment]
When I tried your fix I saw anomalies in more complex use cases, sadly.

The following modification in your Examples notebook freezes the kernel if you run the cell twice, for example.  The button doesn't appear the second time, and "interrupt kernel" doesn't work.

```Python

t0 = time.time()
xx = ["-_-", "o_o"]


def on_poll():
    """This is called repeatedly by run_ui_poll_loop

    Return None if condition hasn't been met yet

    Return some result once done, in this example result
    is a number of seconds it took to press the button 10 times.
    """
    if int(btn.description) < 10:
        print(xx[0], end="\r", flush=True)
        xx[:] = xx[::-1]
        return None  # Continue polling

    # Terminate polling and return final result
    return time.time() - t0

btn = test_button()
print("Press button 10 times")
display(btn)

dt = await run_ui_poll_loop(on_poll, 1 / 15)
print("XXX")  # This should display the text in the output of this cell
btn.description = "0"
dt = await run_ui_poll_loop(on_poll, 1 / 15)
print("._.")  # This should display the text in the output of this cell
n_times = "10 times"  # To verify that the rest of this cell executes before executing cells below
```

Here is what it looks like

<img width="992" alt="async anomaly" src="https://user-images.githubusercontent.com/13857252/133524951-c5106df1-1551-4517-acf0-57881e8746b2.png">


Thanks for your efforts!

**Comment by Kirill888** [comment]
I can explain why that is, and this problem actually exists in older version too if you have `await` statements in your cells and use Run All. 

The last thing `run_ui_poll_loop` does before returning execution control back to calling code is schedule execution of delayed cells using "asyncio.ensure_future" on a current event loop. In a synchronous cell this only gets to run when cell execution is finished, but if you have `await` calls between `run_ui_poll_loop` and end of cell other cells will start executing before current cell is done. Ideally rerunning delayed cells should only be scheduled just after current cell execution is complete, not sure if there are ipykernel hooks for that.

**Comment by AaronWatters** [comment]
Thanks for the explanation.

Unless there is some way around this I need some other approach, unfortunately.

Please see my comment at the bottom here: https://github.com/ipython/ipykernel/issues/758

**Comment by AaronWatters** [comment]
FYI, I'm currently moving forward with this work around, which requires all polling to take place in a sub-task.

https://gist.github.com/AaronWatters/acf26623d19245ba4656a4368f6aa546

**Comment by Kirill888** [comment]
@AaronWatters pretty sure I fixed those issues now, also supporting older versions of ipykernel at the same time as new, keeping all the interfaces as before but also adding `async with` and `async for`.

**Comment by AaronWatters** [comment]
This is a really excellent development especially since it seems like I don't have to
change my client libraries at all!

I thought at one point I saw an anomaly, but apparently it is not easy to reproduce
if it wasn't a mirage.

Thanks very much!  I will let you know if I see any problems.  If not I will close the issues
I opened in ipykernel and ipywidgets in a week or so.
